VariantShredding.md

```diff
@@ -25,290 +25,320 @@
-Similarly, for the query `select * from tbl where variant_get(variant_col, ‘$.id’, ‘integer’) = 123`, the scan could first decode the shredded `id` column, and only fetch/decode the full Variant value for rows that pass the filter.
+Variant metadata is stored in the top-level Variant group in a binary `metadata` column regardless of whether the Variant value is shredded.
 
-# Parquet Example
+All `value` columns within the Variant must use the same `metadata`.
+All field names of a Variant, whether shredded or not, must be present in the metadata.
 
-Consider the following Parquet schema together with how Variant values might be mapped to it.
-Notice that we represent each shredded field in `object` as a group of two fields, `typed_value` and `variant_value`.
-We extract all homogenous data items of a certain path into `typed_value`, and set aside incompatible data items in `variant_value`.
-Intuitively, incompatibilities within the same path may occur because we store the shredding schema per Parquet file, and each file can contain several row groups.
-Selecting a type for each field that is acceptable for all rows would be impractical because it would require buffering the contents of an entire file before writing.
+## Value Shredding
 
-Typically, the expectation is that `variant_value` exists at every level as an option, along with one of `object`, `array` or `typed_value`.
-If the actual Variant value contains a type that does not match the provided schema, it is stored in `variant_value`.
-An `variant_value` may also be populated if an object can be partially represented: any fields that are present in the schema must be written to those fields, and any missing fields are written to `variant_value`.
-
-The `metadata` column is unchanged from its unshredded representation, and may be referenced in `variant_value` fields in the shredded data.
+Variant values are stored in Parquet fields named `value`.
+Each `value` field may have an associated shredded field named `typed_value` that stores the value when it matches a specific type.
+When `typed_value` is present, readers **must** reconstruct shredded values according to this specification.
 
+For example, a Variant field, `measurement` may be shredded as long values by adding `typed_value` with type `int64`:
 ```
-optional group variant_col {
- required binary metadata;
- optional binary variant_value;
- optional group object {
-  optional group a {
-   optional binary variant_value;
-   optional int64 typed_value;
-  }
-  optional group b {
-   optional binary variant_value;
-   optional group object {
-    optional group c {
-      optional binary variant_value;
-      optional binary typed_value (STRING);
-    }
-   }
-  }
- }
+required group measurement (VARIANT) {
+  required binary metadata;
+  optional binary value;
```

Comment:
Hi @rdblue , I have thought of a specific scenario: suppose the json is a struct and has 4000 kvs, of which 100 keys intend to construct `typed_value`. The remaining 3900 columns will not be shredded. In this case, they will all be stored in `value`.

Is it possible for us to define the max number of keys contained in one binary `value`, such as in this case, storing a maximum of 1000, so that v1, v2, v3, v4 can be generated to reduce scan IO.

Futher more, we can define the concept of `group` and stipulate that some certain keys are divided into one binary `value1`.

Reply:
@Zouxxyy could you expand on this use-case?  Is there a specific query pattern for this type of division would improve performance for?

Reply:
@emkornfield For the cols that we do not intend to shred, if we categorize them into four groups and happen to query fields within one of those groups, then our scan I/O can still be reduced by three-quarters. 
Of course, I feel that the concept of `group` will make shredding more complex (currently, it is concise and clear enough), I'm just looking to explore the possibilities and hear opinions.